OneSignalSDK/unittest/src/test/java/com/test/onesignal/GenerateNotificationRunner.java

```diff
@@ -1553,6 +1554,167 @@ public void notificationProcessing(Context context, OSNotificationReceived notif
+      OneSignal.setAppContext(ApplicationProvider.getApplicationContext());
+      OneSignal_setupNotificationExtensionServiceClass();
+
+      // 3. Receive a notification
+      Bundle bundle = getBaseNotifBundle();
+      FCMBroadcastReceiver_processBundle(blankActivity, bundle);
+      threadAndTaskWait();
+
+      // 4. Make sure service was called
+      assertNotNull(lastNotificationReceived);
+
+      // 5. Make sure 1 notification exists in DB
+      assertNotificationDbRecords(1);
+   }
+
+   @Test
+   @Config(shadows = { ShadowGenerateNotification.class })
+   public void testNotificationProcessingAndForegroundHandler_completeNotCalled_callsForegroundHandler() throws Exception {
+      // 1. Setup correct notification extension service class
+      startNotificationExtensionService("com.test.onesignal.GenerateNotificationRunner$" +
+              "NotificationExtensionService_notificationProcessingCompleteNotCalled");
+
+      // 2. Init OneSignal
+      OneSignal.setAppId("b2f7f966-d8cc-11e4-bed1-df8f05be55ba");
+      OneSignal.setAppContext(blankActivity);
+      OneSignal.setNotificationWillShowInForegroundHandler(new OneSignal.AppNotificationWillShowInForegroundHandler() {
+         @Override
+         public void notificationWillShowInForeground(OSNotificationGenerationJob.AppNotificationGenerationJob notificationJob) {
+            lastAppNotificationJob = notificationJob;
+            // Call complete to end without waiting default 30 second timeout
+            notificationJob.complete();
+         }
+      });
+      threadAndTaskWait();
+
+      blankActivityController.resume();
+      threadAndTaskWait();
+
+      // 3. Receive a notification in foreground
+      FCMBroadcastReceiver_processBundle(blankActivity, getBaseNotifBundle());
+      threadAndTaskWait();
+
+      // 4. Make sure service was called
+      assertNotNull(lastNotificationReceived);
+
+      // 5. Make sure the AppNotificationJob is not null
+      assertNotNull(lastAppNotificationJob);
+      assertEquals(OneSignal.OSNotificationDisplay.NOTIFICATION, lastAppNotificationJob.getNotificationDisplayOption());
+
+      // 6. Make sure 1 notification exists in DB
+      assertNotificationDbRecords(1);
+   }
+
+   /**
+    * @see #testNotificationProcessing_completeNotCalled
+    * @see #testNotificationProcessingAndForegroundHandler_completeNotCalled_callsForegroundHandler
+    */
+   public static class NotificationExtensionService_notificationProcessingCompleteNotCalled implements OneSignal.NotificationProcessingHandler {
+
+      @Override
+      public void notificationProcessing(Context context, OSNotificationReceived notification) {
+         lastNotificationReceived = notification;
+
+         // Display called to show notification
+         OSNotificationDisplayedResult notificationDisplayedResult = notification.display();
+
+         // Complete not called to end NotificationProcessingHandler, depend on timeout to finish
```

Comment:
Should have `threadAndTaskWait();` or something here to wait for the timeout to fire.

Reply:
We already have the 

```
FCMBroadcastReceiver_processBundle(blankActivity, getBaseNotifBundle());
threadAndTaskWait();
```

that threadAndTaskWait should cover the timeout fire 🤔

Reply:
ah sorry my bad, this is a class to be used by the two tests above. Ignore my comment here.